clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -69,26 +96,53 @@
     private NetworkReceive netInBuffer;
     private NetworkSend netOutBuffer;
 
-    public SaslServerAuthenticator(String node, final Subject subject, KerberosShortNamer kerberosNameParser, int maxReceiveSize) throws IOException {
+    public SaslServerAuthenticator(String node, final Subject subject, KerberosShortNamer kerberosNameParser, String host, int maxReceiveSize) throws IOException {
         if (subject == null)
             throw new IllegalArgumentException("subject cannot be null");
-        if (subject.getPrincipals().isEmpty())
-            throw new IllegalArgumentException("subject must have at least one principal");
         this.node = node;
         this.subject = subject;
         this.kerberosNamer = kerberosNameParser;
         this.maxReceiveSize = maxReceiveSize;
-        saslServer = createSaslServer();
+        this.host = host;
     }
 
     public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder, Map<String, ?> configs) {
         this.transportLayer = transportLayer;
+        this.configs = configs;
+        try {
+            callbackHandler = new SaslServerCallbackHandler(Configuration.getConfiguration(), kerberosNamer);
+            List<String> enabledMechanisms = (List<String>) this.configs.get(SaslConfigs.SASL_ENABLED_MECHANISMS);
+            if (enabledMechanisms == null || enabledMechanisms.isEmpty())
+                this.enabledMechanisms.add((String) this.configs.get(SaslConfigs.SASL_MECHANISM));
+            else
+                this.enabledMechanisms.addAll(enabledMechanisms);
+        } catch (IOException e) {
+            throw new KafkaException(e);
+        }
     }
 
-    private SaslServer createSaslServer() throws IOException {
+    private void createSaslServer(String mechanism) throws IOException {
+        this.saslMechanism = mechanism;
+        callbackHandler.configure(configs, Mode.SERVER, subject, saslMechanism);
```

Comment:
@ijuma It was written this way when callback handler class was configurable in the original KIP. Since it is no longer configurable, I have moved the construction to `createSaslServer` when the mechanism is known.